node/src/chain/fsm/outofsync.rs

```diff
@@ -176,32 +177,35 @@ impl<DB: database::DB, VM: vm::VMExecution, N: Network>
             attempts: 3,
         }
     }
+
     /// Performed when entering the OutOfSync state
     ///
     /// Handles the logic for entering the out-of-sync state. Sets the target
     /// block range, adds the `target_block` to the pool, updates the
     /// `remote_peer` address, and starts to request missing blocks
-    pub async fn on_entering(
-        &mut self,
-        target_block: Block,
-        peer_addr: SocketAddr,
-    ) {
+    pub async fn on_entering(&mut self, presync: PresyncInfo) {
+        let peer_addr = presync.peer_addr;
+        let pool = presync.pool;
         let curr_height = self.acc.read().await.get_curr_height().await;
 
-        self.range = (curr_height + 1, target_block.header().height);
+        self.range = (curr_height + 1, presync.remote_height);
 
         // add target_block to the pool
-        let key = target_block.header().height;
         self.drain_pool().await;
-        self.pool.insert(key, target_block);
+        for b in &pool {
+            self.pool.insert(b.header().height, b.clone());
+        }
```

Comment:
I think this is not necessary. If the block needs to be added to the pool (bc it can't be acceoted yet) it will be handled inside `on_block_event`.

Reply:
It's required, otherwise if we received intermediate blocks during the presync, we're going to discard them

And if we discard them, even if we're going to request them again, they won't be received due to network dupemap, leading the outofsync timer to expire

Reply:
Now I got what you meant
Don't add them to self.pool, but directly call the on_block_event against the presync.pool

I thought about this, but I decided to not do it because the on_block_event can request missing blocks according to the self.pool

So I preferred to keep things separate

Reply:
we should probably skip the request if last_request=0 (ie, we did not entered oos yet).
Either way, the whole logic of the presync's pool is only needed because of the getblocks request at the beginning.
In the presync process we only request tip+1 so we shouldn't receive any other block.

Anyway, as it is now it still works, so I don't mind discussing this in a followup pr